packages/core/base/src/signer.ts

```diff
@@ -1,42 +1,87 @@
-import type { Connection, Transaction, TransactionSignature } from '@solana/web3.js';
-import type { SendTransactionOptions, WalletAdapter } from './adapter.js';
+import type { Connection, TransactionSignature, TransactionVersion } from '@solana/web3.js';
+import type { SendTransactionOptions, WalletAdapter, WalletAdapterProps } from './adapter.js';
 import { BaseWalletAdapter } from './adapter.js';
 import { WalletSendTransactionError, WalletSignTransactionError } from './errors.js';
+import type { TransactionOrVersionedTransaction } from './types.js';
 
-export interface SignerWalletAdapterProps {
-    signTransaction(transaction: Transaction): Promise<Transaction>;
-    signAllTransactions(transaction: Transaction[]): Promise<Transaction[]>;
+export interface SignerWalletAdapterProps<
+    Name extends string = string,
+    SupportedTransactionVersions extends Set<TransactionVersion> | null = Set<TransactionVersion> | null
+> extends WalletAdapterProps<Name, SupportedTransactionVersions> {
+    signTransaction(
+        transaction: TransactionOrVersionedTransaction<SupportedTransactionVersions>
+    ): Promise<TransactionOrVersionedTransaction<SupportedTransactionVersions>>;
+    signAllTransactions(
+        transactions: TransactionOrVersionedTransaction<SupportedTransactionVersions>[]
+    ): Promise<TransactionOrVersionedTransaction<SupportedTransactionVersions>[]>;
```

Comment:
These types are not perfectly ideal. Let's say you have
```ts
const tx = new Transaction();
const signedTx = await adapter.signTransaction(tx);
```

In this case, if `adapter.supportedTransactionVersions` is `null`, you know `signedTx` _must be_ a `Transaction`, but its actual type is either `Transaction` or `Transaction | VersionedTransaction`.

If we declared it this way, it solves this problem:
```ts
<T extends TransactionOrVersionedTransaction<SupportedTransactionVersions>>signTransaction(transaction: T): Promise<T>;
```
But doing so makes the implementation of the classes that use this interface require a bunch of type wrangling.

Reply:
This is fixed now.